trie/snap_disklayer.go

```diff
@@ -0,0 +1,247 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/VictoriaMetrics/fastcache"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// diskLayer is a low level persistent snapshot built on top of a key-value store.
+type diskLayer struct {
+	root   common.Hash // Immutable, root hash of the base snapshot
+	diffid uint64      // Immutable, corresponding reverse diff id
+
+	diskdb ethdb.Database   // Key-value store containing the base snapshot
+	clean  *fastcache.Cache // Clean node cache to avoid hitting the disk for direct access
+	dirty  *diskcache       // Dirty node cache to aggregate writes and temporary cache.
+	stale  bool             // Signals that the layer became stale (state progressed)
+	lock   sync.RWMutex     // Lock used to protect stale flag
+}
+
+// newDiskLayer creates a new disk layer based on the passing arguments.
+func newDiskLayer(root common.Hash, diffid uint64, clean *fastcache.Cache, dirty *diskcache, diskdb ethdb.Database) *diskLayer {
+	return &diskLayer{
+		diskdb: diskdb,
+		clean:  clean,
+		dirty:  dirty,
+		root:   root,
+		diffid: diffid,
+	}
+}
+
+// Root returns root hash of corresponding state.
+func (dl *diskLayer) Root() common.Hash {
+	return dl.root
+}
+
+// Parent always returns nil as there's no layer below the disk.
+func (dl *diskLayer) Parent() snapshot {
+	return nil
+}
+
+// ID returns the id of associated reverse diff.
+func (dl *diskLayer) ID() uint64 {
+	return dl.diffid
+}
+
+// Stale return whether this layer has become stale (was flattened across) or if
+// it's still live.
+func (dl *diskLayer) Stale() bool {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	return dl.stale
+}
+
+// MarkStale sets the stale flag as true.
+func (dl *diskLayer) MarkStale() {
+	dl.lock.Lock()
+	defer dl.lock.Unlock()
+
+	if dl.stale == true {
+		panic("triedb disk layer is stale") // we've committed into the same base from two children, boom
+	}
+	dl.stale = true
+}
+
+// Node retrieves the trie node associated with a particular key.
+func (dl *diskLayer) Node(storage []byte, hash common.Hash) (*cachedNode, error) {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	if dl.stale {
+		return nil, errSnapshotStale
+	}
+	// Try to retrieve the trie node from the dirty memory cache.
+	// The map is lock free since it's impossible to mutate the
+	// disk layer before tagging it as stale.
+	n, err := dl.dirty.node(storage, hash)
+	if err != nil {
+		return nil, err
+	}
+	if n != nil {
+		return n, nil
```

Comment:
Might make sense to move 

```
        triedbDirtyHitMeter.Mark(1)
        triedbDirtyNodeHitDepthHist.Update(int64(128))
        triedbDirtyReadMeter.Mark(int64(n.size))
```

here so that the accounting is done the same way for both hits/misses and dirty/cleans,